build/rollup.ts

```diff
@@ -184,6 +184,14 @@ async function buildUserscriptPassTwo(passOneResult: Readonly<RollupOutput>, use
                 branchName: 'dist',
                 include: /index\.js/,
             }),
+            {
+                name: 'ProperIIFE',
+                // Need to bind `this` in the invocation of the IIFE since it's
+                // used by the GM4 polyfill.
+                renderChunk(code: string): string {
+                    return code.replace(/}\)\(\);(\s*)$/, '}).bind(this)();$1');
+                },
+            },
```

Comment:
I just wanted to test what happens when I remove this code.
In which cases is it actually necessary to have this patch? I had assumed that only TM needs the polyfill, but I have just tested with GM, TM and VM in Firefox 94 and it seemingly had no effect whether the userscript binds `this` or not. Next step would be testing without the entire IIFE but since I could not trigger a bug or a warning, I have not proceeded to this step yet.

Reply:
As far as I could tell, all of the userscript engines set `this` to either the `window` or the "context" in which the script runs (GM's isolated content context), but of course in an IIFE that isn't a lambda, the top-level `this` is lost if we don't bind it. We need the top-level `this` because `gm4-polyfill` does `this.GM = {}` and looks up e.g. `GM_xmlhttpRequest` as `this["GM_xmlhttpRequest"]`. If it's undefined, that doesn't work. However, the polyfill likely doesn't do this if the userscript engine already sets `GM.*` functions, but I'd have to test it more myself.

I'd be entirely in favour of getting rid of the IIFE. However, I'm not 100% sure we actually can, since rollup doesn't really have a "bare" output format. CJS output puts `require` statements for the `vendor` and `lib` chunks, so I'm guessing ESM will also inject `imports`. I'll investigate whether we can do something about that.

To be honest, I'd also like to drop `gm4-polyfill` and roll our own solution which we can tree-shake so we don't polyfill functions which we don't use, and without relying on `this`.

Reply:
It's weird that you don't run into issues when you don't bind `this`. I get:
```
Script error in [Greasemonkey script https://github.com/ROpdebee/mb-userscripts/MB: Enhanced Cover Art Uploads; version 2021.11.7]:
TypeError: can't access property "GM_addStyle", _this is undefined
```
(the minified code at some point does `_this = this`).

Reply:
I just found out that I wasn't running into any issues simply because the latest version of the userscript, which I had built on my PC, didn't even include the GM polyfill, see my latest comment above 😨

Now that the polyfill is included, I am also seeing the same error if I am not binding `this` or removing the whole outer IIFE by hand 😅

But the fact that I was able to succesfully test the userscript without the polyfill in all three userscript engines also means that we probably don't need the entire polyfill at all, unless we want to support the deprecated GM v3. I am surprised that TM also supports the new `GM.*` functions although that's not documented on the official website (I mainly used the `GM_*` functions because I assumed that TM supports only those).

Reply:
I'd be fine with just adding a `compat.ts` module which exports the correct function depending on which one is available. E.g. `export const GMxmlhttpRequest = (typeof GM_xmlhttpRequest !== 'undefined') ? GM_xmlhttpRequest : GM.xmlHttpRequest` instead of using the polyfill. That way we don't include polyfills for stuff we don't need and we don't need any of these hacks either.

Reply:
I would also be in favour of using a custom compatibility module.
In addition to the reasons you gave above, I just found out that we don't have full compatibility with the GM v4 API in the latest versions of GM, VM and TM after all:
Testing the userscript variant without the polyfill, I was just blessed with the following error for a release which is linked to Amazon: `Uncaught (in promise) TypeError: GM.getResourceUrl is not a function`
Apparently the method is called `getResourceURL` in [Violentmonkey's API](https://violentmonkey.github.io/api/gm/#gm) 🤦

Reply:
I've added the `compat` module and reverted the polyfill changes and hacks, tested on all the browsers (including Waterfox under GM3) and it still works for me.